guides/api-style.rst

```diff
@@ -0,0 +1,125 @@
+* Exception to the above: Strategies for collection types should *not* have a
+  default argument for element strategies.
+* Interacting arguments (e.g. arguments that must be in a particular order, or
+  where at most one is valid, or where one argument restricts the valid range
+  of the other) are fine, but when this happens the behaviour of defaults
+  should automatically be adjusted. e.g. if the normal default of an argument
+  would become invalid, the function should still do the right thing if that
+  default is used.
+* Where the actual default used depends on other arguments, the default parameter
+  should be None.
+* It's worth thinking about the order of arguments: the first one or two
+  arguments are likely to be passed positionally, so try to put values there
+  where this is useful and not too confusing.
+* Arguments should not be "a value or a strategy for generating that value".
+  If you find yourself inclined to write something like that, instead make it
+  take a strategy. If a user wants to pass a value they can wrap it in a call
+  to ``just``.
+* When adding arguments to strategies, think carefully about whether the user
+  is likely to want that value to vary often. If so, make it a strategy instead
+  of a value. In particular if it's likely to be common that they would want to
+  write ``some_strategy.flatmap(lambda x: my_new_strategy(argument=x))`` then
+  it should be a strategy.
+
+~~~~~~~~~~~~~~
+Function Names
+~~~~~~~~~~~~~~
+
+We don't have any real consistency here. The rough approach we follow is:
+
+* Names are `snake_case` as is standard in Python.
+* Strategies for a particular type are typically named as a plural name for
+  that type. Where that type has some truncated form (e.g. int, str) we use a
+  longer form name.
+* Other strategies have no particular common naming convention.
+
+~~~~~~~~~~~~~~
+Argument Names
+~~~~~~~~~~~~~~
+
+We should try to use the same argument names and orders across different
+strategies wherever possible. In particular:
+
+* For collection types, the element strategy (or strategies) should always be
+  the first argument. It does not have a specific common name.
+* For ordered types, the first two arguments should be a lower and an upper
+  bound. They should be called ``min_value`` and ``max_value``.
+* Collection types should have a ``min_size`` and a ``max_size`` parameter that
+  controls the range of their size. ``min_size`` should default to zero and
+  ``max_size`` to ``None`` (even if internally it is bounded).
+
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+A catalogue of current violations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following are places where we currently deviate from this style. Some of
+these should be considered targets for deprecation and/or improvement.
+
+* most of the collections in ``hypothesis.strategies`` have an ``average_size``
+  distribution hint.
+* many of the collections in ``hypothesis.strategies`` allow a default of
```

Comment:
Above, we say that "Strategies for collection types should *not* have a default argument for element strategies." - I'd keep that and delete this, or clarify that you mean an elements strategy should be a required argument and empty collections provided by `builds(collection_type)`.

Reply:
This is in the exceptions section specifically as a thing we do but shouldn't!

Reply:
For clarity then, I'd also mention what should be done instead.

Reply:
To clarify then, "Instead, `elements` should be a required argument and empty collections constructed with `builds()`."

Reply:
I really don't think it's unclear. It's in a section explicitly labelled violations and the resolution is "Don't do that then" the same as all the other violations.